pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -45,127 +51,212 @@ type Reporter interface {
-			PipedId:              app.PipedId,
-			ProjectId:            app.ProjectId,
-			Kind:                 app.Kind,
-			ApplicationLiveState: res.GetApplicationLiveState(),
+	appsPerReporter := len(apps) / limit
+	if appsPerReporter == 0 {
+		appsPerReporter = 1
+	}
+
+	appsGrouped := make([][]*model.Application, 0)
+	for i := 0; i < len(apps); i += appsPerReporter {
+		end := i + appsPerReporter
+		if end > len(apps) {
+			end = len(apps)
 		}
-		snapshot.DetermineApplicationHealthStatus()
-
-		if _, err := pr.apiClient.ReportApplicationLiveState(ctx, &pipedservice.ReportApplicationLiveStateRequest{
-			Snapshot: snapshot,
-		}); err != nil {
-			pr.logger.Error("failed to report application live state",
-				zap.String("application-id", app.Id),
-				zap.Error(err),
-			)
+		appsGrouped = append(appsGrouped, apps[i:end])
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(limit)
+
+	r.logger.Info("start flushing snapshots", zap.Int("application-count", len(apps)), zap.Int("concurrency", limit))
+	for _, apps := range appsGrouped {
+		eg.Go(func() error {
+			r.flushAll(ctx, apps, repoMap)
+			return nil
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		r.logger.Error("failed to flush snapshots", zap.Error(err))
+	}
+
+	r.logger.Info("successfully flushed snapshots")
+}
+
+func (r *reporter) flushAll(ctx context.Context, apps []*model.Application, repoMap map[string]git.Repo) {
+	for _, app := range apps {
+		repo, ok := repoMap[app.GitPath.Repo.Id]
+		if !ok {
+			r.logger.Error("failed to find repository for application", zap.String("application-id", app.Id))
 			continue
 		}
 
-		// Report the application sync state to the control plane.
-		if _, err := pr.apiClient.ReportApplicationSyncState(ctx, &pipedservice.ReportApplicationSyncStateRequest{
-			ApplicationId: app.Id,
-			State:         res.GetSyncState(),
-		}); err != nil {
-			pr.logger.Error("failed to report application live state",
-				zap.String("application-id", app.Id),
-				zap.Error(err),
-			)
+		if err := r.flush(ctx, app, repo); err != nil {
+			r.logger.Error("failed to flush application live state", zap.String("application-id", app.Id), zap.Error(err))
 			continue
 		}
+	}
+}
+
+func (r *reporter) flush(ctx context.Context, app *model.Application, repo git.Repo) error {
+	dir, err := os.MkdirTemp(r.workingDir, fmt.Sprintf("app-%s-*", app.Id))
+	if err != nil {
+		r.logger.Error("failed to create temporary directory", zap.Error(err))
+		return err
+	}
+
+	dsp := deploysource.NewProvider(dir, deploysource.NewLocalSourceCloner(repo, "target", "HEAD"), app.GitPath, r.secretDecrypter)
+	ds, err := dsp.Get(ctx, io.Discard)
+	if err != nil {
+		r.logger.Error("failed to get deploy source", zap.String("application-id", app.Id), zap.Error(err))
+		return err
+	}
+
+	cfg, err := config.DecodeYAML[*config.GenericApplicationSpec](ds.ApplicationConfig)
+	if err != nil {
+		r.logger.Error("unable to parse application config", zap.Error(err))
+		return err
+	}
```

Comment:
We use the Deploy Source for two reasons.

1. to use app.pipecd.yaml for the app
2. to pass it when calling `GetLivestate`. It expects to use it do do drift detection.